File: fe/src/lib/api/inventoryService.ts
import { apiClient } from "./apiClient";

export interface Asset {
  id: string;
  name: string;
  code: string;
  category: string;
  description: string;
  min_stock: number;
  unit: string;
  created_at: string;
  updated_at: string;
  stock_summary?: StockSummary;
}

export interface GlobalSummary {
  total_assets: number;
  active_items: number;
  low_stock_assets: number;
}

export interface StockSummary {
  total: number;
  in_stock: number;
  deployed: number;
  maintenance: number;
  low_stock: boolean;
}

export interface AssetInstance {
  id: string;
  asset_id: string;
  serial_number: string;
  status: 'in_stock' | 'deployed' | 'maintenance' | 'disposed' | 'sold';
  condition: 'new' | 'second' | 'broken' | 'refurbished';
  location: string;
  last_checked_at?: string;
  last_checked_by?: string;
  created_at: string;
}

export interface AssetLog {
  id: string;
  action: string;
  from_value: string;
  to_value: string;
  actor: string;
  notes: string;
  created_at: string;
}

export interface CreateAssetRequest extends Partial<Asset> {
  initial_stock?: number;
  initial_condition?: string;
}

export const inventoryService = {
  listAssets: (params?: { search?: string; category?: string; page?: number; page_size?: number }) =>
    apiClient.get<{ data: Asset[]; total: number }>("/inventory/assets", { params }),

  getAsset: (id: string) =>
    apiClient.get<Asset>(`/inventory/assets/${id}`),

  getSummary: () =>
    apiClient.get<GlobalSummary>(`/inventory/summary`),

  createAsset: (data: CreateAssetRequest) => {
    const { initial_stock, initial_condition, ...assetData } = data;
    return apiClient.post<Asset>("/inventory/assets", {
      asset: assetData,
      initial_stock,
      initial_condition
    });
  },

  listInstances: (assetId: string) =>
    apiClient.get<{ data: AssetInstance[] }>(`/inventory/assets/${assetId}/instances`),

  addInstance: (assetId: string, data: Partial<AssetInstance>) =>
    apiClient.post<AssetInstance>(`/inventory/assets/${assetId}/instances`, data),

  updateInstance: (assetId: string, id: string, data: Partial<AssetInstance>) =>
    apiClient.put(`/inventory/assets/${assetId}/instances/${id}`, data),

  bulkUpdate: (assetId: string, status: string) =>
    apiClient.post(`/inventory/assets/${assetId}/bulk-update`, { status }),

  getHistory: (params: { asset_id?: string; instance_id?: string }) =>
    apiClient.get<{ data: AssetLog[] }>("/inventory/history", { params }),

  deleteAsset: (id: string) =>
    apiClient.delete(`/inventory/assets/${id}`),

  getPublicInstance: (id: string) =>
    apiClient.get<AssetInstance & { asset: Asset }>(`/public/inventory/instance/${id}`),
};

File: BE/internal/service/inventory_service.go
package service

import (
	"context"
	"fmt"

	"rrnet/internal/domain/inventory"
	"rrnet/internal/repository"

	"github.com/google/uuid"
)

type InventoryService struct {
	repo *repository.InventoryRepository
}

func NewInventoryService(repo *repository.InventoryRepository) *InventoryService {
	return &InventoryService{repo: repo}
}

// --- Asset Logic ---

func (s *InventoryService) CreateAsset(ctx context.Context, tenantID uuid.UUID, a *inventory.Asset, initialStock int, condition inventory.InstanceCondition, actor string) error {
	a.ID = uuid.New()
	a.TenantID = tenantID

	if err := s.repo.CreateAsset(ctx, a); err != nil {
		return err
	}

	// Create initial instances if stock > 0
	for i := 0; i < initialStock; i++ {
		inst := &inventory.AssetInstance{
			AssetID:      a.ID,
			SerialNumber: fmt.Sprintf("%s-%03d", a.Code, i+1),
			Status:       inventory.StatusInStock,
			Condition:    condition,
			Location:     "Warehouse / Central",
		}
		_ = s.AddInstance(ctx, tenantID, inst, actor)
	}

	return nil
}

func (s *InventoryService) GetAsset(ctx context.Context, tenantID, assetID uuid.UUID) (*inventory.Asset, error) {
	return s.repo.GetAssetByID(ctx, tenantID, assetID)
}

func (s *InventoryService) ListAssets(ctx context.Context, tenantID uuid.UUID, filter *inventory.AssetFilter) ([]*inventory.Asset, int, error) {
	return s.repo.ListAssets(ctx, tenantID, filter)
}

func (s *InventoryService) GetGlobalSummary(ctx context.Context, tenantID uuid.UUID) (*inventory.GlobalSummary, error) {
	return s.repo.GetGlobalSummary(ctx, tenantID)
}

func (s *InventoryService) DeleteAsset(ctx context.Context, tenantID, assetID uuid.UUID, actor string) error {
	err := s.repo.DeleteAsset(ctx, tenantID, assetID)
	if err == nil {
		_ = s.repo.CreateLog(ctx, &inventory.AssetLog{
			TenantID: tenantID,
			AssetID:  &assetID,
			Action:   "asset_deleted",
			Actor:    actor,
			Notes:    fmt.Sprintf("Asset and all units soft-deleted by %s", actor),
		})
	}
	return err
}

// --- Instance Logic ---

func (s *InventoryService) AddInstance(ctx context.Context, tenantID uuid.UUID, inst *inventory.AssetInstance, actor string) error {
	inst.ID = uuid.New()
	inst.TenantID = tenantID

	// Generate serial number if empty
	if inst.SerialNumber == "" {
		asset, err := s.repo.GetAssetByID(ctx, tenantID, inst.AssetID)
		if err == nil {
			count, _ := s.repo.CountInstances(ctx, tenantID, inst.AssetID)
			inst.SerialNumber = fmt.Sprintf("%s-%03d", asset.Code, count+1)
		}
	}

	err := s.repo.CreateInstance(ctx, inst)
	if err != nil {
		return err
	}

	// Log the addition
	return s.repo.CreateLog(ctx, &inventory.AssetLog{
		TenantID:   tenantID,
		AssetID:    &inst.AssetID,
		InstanceID: &inst.ID,
		Action:     "CHECK_IN",
		ToValue:    string(inst.Status),
		Actor:      actor,
		Notes:      fmt.Sprintf("Initial check-in at %s", inst.Location),
	})
}

func (s *InventoryService) ListInstances(ctx context.Context, tenantID uuid.UUID, assetID uuid.UUID) ([]*inventory.AssetInstance, error) {
	return s.repo.ListInstances(ctx, tenantID, assetID)
}

func (s *InventoryService) GetInstanceDetail(ctx context.Context, id uuid.UUID) (*inventory.AssetInstance, error) {
	return s.repo.GetInstanceByID(ctx, id)
}

func (s *InventoryService) UpdateInstance(ctx context.Context, tenantID uuid.UUID, assetID uuid.UUID, id uuid.UUID, req *inventory.AssetInstance, actor string) error {
	// Get current state for logging
	current, err := s.repo.ListInstances(ctx, tenantID, assetID)
	if err != nil {
		return err
	}

	var target *inventory.AssetInstance
	for _, i := range current {
		if i.ID == id {
			target = i
			break
		}
	}

	if target == nil {
		return repository.ErrInstanceNotFound
	}

	// Update record
	target.Status = req.Status
	target.Condition = req.Condition
	target.Location = req.Location
	target.LastCheckedBy = &actor

	err = s.repo.UpdateInstance(ctx, target)
	if err != nil {
		return err
	}

	// Log change
	return s.repo.CreateLog(ctx, &inventory.AssetLog{
		TenantID:   tenantID,
		AssetID:    &assetID,
		InstanceID: &id,
		Action:     "STATUS_UPDATE",
		FromValue:  string(target.Status), // Simplified for log
		ToValue:    string(req.Status),
		Actor:      actor,
		Notes:      fmt.Sprintf("Condition: %s, Location: %s", req.Condition, req.Location),
	})
}

// --- Bulk Logic ---

func (s *InventoryService) BulkStatusUpdate(ctx context.Context, tenantID uuid.UUID, assetID uuid.UUID, status inventory.InstanceStatus, actor string) error {
	instances, err := s.repo.ListInstances(ctx, tenantID, assetID)
	if err != nil {
		return err
	}

	for _, inst := range instances {
		if inst.Status == status {
			continue
		}

		oldStatus := inst.Status
		inst.Status = status
		inst.LastCheckedBy = &actor

		_ = s.repo.UpdateInstance(ctx, inst)

		// Log each
		_ = s.repo.CreateLog(ctx, &inventory.AssetLog{
			TenantID:   tenantID,
			AssetID:    &assetID,
			InstanceID: &inst.ID,
			Action:     "BULK_STATUS_CHANGE",
			FromValue:  string(oldStatus),
			ToValue:    string(status),
			Actor:      actor,
			Notes:      "Applied via global settings",
		})
	}

	return nil
}

// --- Audit Logic ---

func (s *InventoryService) GetHistory(ctx context.Context, tenantID uuid.UUID, assetID *uuid.UUID, instanceID *uuid.UUID) ([]*inventory.AssetLog, error) {
	return s.repo.ListLogs(ctx, tenantID, assetID, instanceID)
}
